fish-rust/src/common.rs

```diff
@@ -1985,6 +1995,100 @@ pub fn get_by_sorted_name<T: Named>(name: &wstr, vals: &'static [T]) -> Option<&
     }
 }
 
+/// Takes ownership of a variable and `Deref`s/`DerefMut`s into a projection of that variable.
+///
+/// Can be used as a workaround for the lack of `MutexGuard::map()` to return a `MutexGuard`
+/// exposing only a variable of the Mutex-owned object.
+pub struct Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    value: T,
+    view: F1,
+    view_mut: F2,
+}
+
+impl<T, V, F1, F2> Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    pub fn new(owned: T, project: F1, project_mut: F2) -> Self {
+        Projection {
+            value: owned,
+            view: project,
+            view_mut: project_mut,
+        }
+    }
+}
+
+impl<T, V, F1, F2> Deref for Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        (self.view)(&self.value)
+    }
+}
+
+impl<T, V, F1, F2> DerefMut for Projection<T, V, F1, F2>
+where
+    F1: Fn(&T) -> &V,
+    F2: Fn(&mut T) -> &mut V,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        (self.view_mut)(&mut self.value)
+    }
+}
+
+/// A trait to make it more convenient to pass ascii/Unicode strings to functions that can take
+/// non-Unicode values.
+pub trait ToCString {
+    fn to_cstring(self) -> CString;
+}
+
+impl ToCString for &str {
+    fn to_cstring(self) -> CString {
+        CString::new(self).unwrap()
+    }
+}
+
+impl ToCString for CString {
+    fn to_cstring(self) -> CString {
+        self
+    }
+}
+
+impl ToCString for &CStr {
+    fn to_cstring(self) -> CString {
+        self.to_owned()
+    }
+}
+
+impl ToCString for &wstr {
+    /// The wide string may contain non-Unicode bytes mapped to the private-use Unicode range, so we
+    /// have to use [`wcs2zstring()`](self::wcs2zstring) to convert it correctly.
+    fn to_cstring(self) -> CString {
+        self::wcs2zstring(self)
+    }
+}
+
+impl ToCString for &Utf32String {
+    fn to_cstring(self) -> CString {
+        self.as_utfstr().to_cstring()
+    }
+}
+
+impl ToCString for &String {
```

Comment:
I'm OK with `ToCString` but we should NOT have an implementation for `&String` or `&str`. There is no guarantee that the user's encoding us UTF8.

Reply:
IMHO that's all the more reason to *keep* that trait impl (but change it) because there should be an easy way to use `&str` or else people will just be `CString::new(s).unwrap()` with the `&str`. 

I'll change the impl to convert to a wide string then use that impl (which goes through `wstr2zstring()` or whatever its called and is safe).

Reply:
I don't like to introduce inconsistency / multiple obvious ways to do the same thing.
If this trait has advantages over a free function, then we should update the entire tree to use the `to_cstring` name and delete `wcs2zstring` including on the C++ side.